src/raftstore/store/peer_storage.rs

```diff
@@ -1282,8 +1282,12 @@ pub fn do_snapshot(
 
     let key = SnapKey::new(region_id, term, idx);
 
-    mgr.register(key.clone(), SnapEntry::Generating);
-    defer!(mgr.deregister(&key, &SnapEntry::Generating));
+    if mgr.register(key.clone(), SnapEntry::Generating).is_err() {
```

Comment:
need to check error type?

Reply:
It only fails when there is already a `Generating`. So I think return `Unavaliable` here is OK.

Reply:
how about adding new error later?  the error handling here confuses me, we must use a clear way.

Reply:
Even if we add new errors later, what we can do here is still return a `Unavaliable`. Other errors will cause raft-rs panic.